File: calculator.py
"""
Simple Calculator in Python
Supports: addition, subtraction, multiplication, division,
          modulus, exponentiation, integer division, average,
          percentage, absolute difference, maximum, and minimum.
"""


def add(a, b):
    return a + b


def subtract(a, b):
    return a - b


def multiply(a, b):
    return a * b


def divide(a, b):
    if b == 0:
        raise ZeroDivisionError("Cannot divide by zero.")
    return a / b


def modulus(a, b):
    if b == 0:
        raise ZeroDivisionError("Cannot perform modulus by zero.")
    return a % b


def power(a, b):
    return a ** b


def floor_divide(a, b):
    if b == 0:
        raise ZeroDivisionError("Cannot perform floor division by zero.")
    return a // b


def average(a, b):
    return (a + b) / 2


def percentage(a, b):
    if b == 0:
        raise ZeroDivisionError("Cannot calculate percentage with zero total.")
    return (a / b) * 100


def absolute_difference(a, b):
    return abs(a - b)


def maximum(a, b):
    return max(a, b)


def minimum(a, b):
    return min(a, b)


OPERATIONS = {
    "1": ("Add (+)", add),
    "2": ("Subtract (-)", subtract),
    "3": ("Multiply (*)", multiply),
    "4": ("Divide (/)", divide),
    "5": ("Modulus (%)", modulus),
    "6": ("Power (**)", power),
    "7": ("Floor Divide (//)", floor_divide),
    "8": ("Average", average),
    "9": ("Percentage (a is what % of b)", percentage),
    "10": ("Absolute Difference (|a - b|)", absolute_difference),
    "11": ("Maximum", maximum),
    "12": ("Minimum", minimum),
}


def get_number(prompt):
    while True:
        try:
            return float(input(prompt))
        except ValueError:
            print("Invalid number. Please try again.")


def show_menu():
    print("\n===== Calculator =====")
    for key, (label, _) in OPERATIONS.items():
        print(f"{key}. {label}")
    print("0. Exit")


def main():
    print("Welcome to the Python Calculator!")
    while True:
        show_menu()
        choice = input("Choose an option: ").strip()

        if choice == "0":
            print("Goodbye!")
            break

        if choice not in OPERATIONS:
            print("Invalid choice. Please select a valid option.")
            continue

        label, func = OPERATIONS[choice]
        num1 = get_number("Enter the first number: ")
        num2 = get_number("Enter the second number: ")

        try:
            result = func(num1, num2)
            print(f"Result of {label}: {result}")
        except ZeroDivisionError as e:
            print(f"Error: {e}")


if __name__ == "__main__":
    main()

File: Calculator.java
import java.util.Scanner;

/**
 * A simple console-based calculator that supports basic arithmetic operations:
 * addition, subtraction, multiplication, division, modulus, power, square root,
 * percentage, and absolute value.
 *
 * Usage: javac Calculator.java && java Calculator
 */
public class Calculator {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        boolean keepRunning = true;

        System.out.println("=================================");
        System.out.println("       Java Calculator");
        System.out.println("=================================");

        while (keepRunning) {
            printMenu();
            String choice = scanner.nextLine().trim();

            switch (choice) {
                case "1":
                case "2":
                case "3":
                case "4":
                case "5":
                case "6":
                case "8":
                    performOperation(scanner, choice);
                    break;
                case "7":
                case "9":
                    performUnaryOperation(scanner, choice);
                    break;
                case "10":
                    keepRunning = false;
                    System.out.println("Goodbye!");
                    break;
                default:
                    System.out.println("Invalid choice. Please select 1-10.");
            }
        }

        scanner.close();
    }

    private static void printMenu() {
        System.out.println();
        System.out.println("Choose an operation:");
        System.out.println("  1. Addition (+)");
        System.out.println("  2. Subtraction (-)");
        System.out.println("  3. Multiplication (*)");
        System.out.println("  4. Division (/)");
        System.out.println("  5. Modulus (%)");
        System.out.println("  6. Power (x^y)");
        System.out.println("  7. Square Root (sqrt x)");
        System.out.println("  8. Percentage (x% of y)");
        System.out.println("  9. Absolute Value (|x|)");
        System.out.println(" 10. Exit");
        System.out.print("Enter your choice: ");
    }

    private static void performOperation(Scanner scanner, String choice) {
        double a = readNumber(scanner, "Enter the first number: ");
        double b = readNumber(scanner, "Enter the second number: ");
        double result;

        switch (choice) {
            case "1":
                result = a + b;
                System.out.printf("Result: %s + %s = %s%n", format(a), format(b), format(result));
                break;
            case "2":
                result = a - b;
                System.out.printf("Result: %s - %s = %s%n", format(a), format(b), format(result));
                break;
            case "3":
                result = a * b;
                System.out.printf("Result: %s * %s = %s%n", format(a), format(b), format(result));
                break;
            case "4":
                if (b == 0) {
                    System.out.println("Error: Division by zero is not allowed.");
                    return;
                }
                result = a / b;
                System.out.printf("Result: %s / %s = %s%n", format(a), format(b), format(result));
                break;
            case "5":
                if (b == 0) {
                    System.out.println("Error: Modulus by zero is not allowed.");
                    return;
                }
                result = a % b;
                System.out.printf("Result: %s %% %s = %s%n", format(a), format(b), format(result));
                break;
            case "6":
                result = Math.pow(a, b);
                if (Double.isNaN(result) || Double.isInfinite(result)) {
                    System.out.println("Error: Result is undefined or out of range.");
                    return;
                }
                System.out.printf("Result: %s ^ %s = %s%n", format(a), format(b), format(result));
                break;
            case "8":
                // Interpreted as: a percent of b  ->  (a / 100) * b
                result = (a / 100.0) * b;
                System.out.printf("Result: %s%% of %s = %s%n", format(a), format(b), format(result));
                break;
            default:
                System.out.println("Unknown operation.");
        }
    }

    private static void performUnaryOperation(Scanner scanner, String choice) {
        double a = readNumber(scanner, "Enter the number: ");
        double result;

        switch (choice) {
            case "7":
                if (a < 0) {
                    System.out.println("Error: Cannot take the square root of a negative number.");
                    return;
                }
                result = Math.sqrt(a);
                System.out.printf("Result: sqrt(%s) = %s%n", format(a), format(result));
                break;
            case "9":
                result = Math.abs(a);
                System.out.printf("Result: |%s| = %s%n", format(a), format(result));
                break;
            default:
                System.out.println("Unknown operation.");
        }
    }

    private static double readNumber(Scanner scanner, String prompt) {
        while (true) {
            System.out.print(prompt);
            String input = scanner.nextLine().trim();
            try {
                return Double.parseDouble(input);
            } catch (NumberFormatException e) {
                System.out.println("Invalid number. Please try again.");
            }
        }
    }

    /**
     * Formats a double so whole numbers display without trailing ".0".
     */
    private static String format(double value) {
        if (value == Math.floor(value) && !Double.isInfinite(value)) {
            return String.valueOf((long) value);
        }
        return String.valueOf(value);
    }
}
